### PR TITLE
Windows compatible version. Not backwards compatible yet. 

### PR DESCRIPTION
**Pull request by tgoossens**
This pull request is not intended for actual merging with release. But rather as a means to ask for feedback and help of the maintainer of smos.

This version has modifications that make 'stack install smos' succeed. Other tools do not yet succeed.

The smos.exe file can be run in powershell. But to make timezones work you first need to set the TZ and TZDIR environment variable:

in powershell (change the paths and time zones accordingly):
$env:TZDIR = "C:\Users\thomas\Desktop\"  
$env:TZ= "Amsterdam"
smos

### CODE REVIEW

**Comment by tgoossens** [comment]
Also, please note that currently you cannot clone smos on windows at the moment because an unacceptable filename for windows:

 
Cloning into 'C:\Users\Thomas\Documents\GitHub\test\smos'...
remote: Enumerating objects: 43138, done.        
remote: Counting objects: 100% (2962/2962), done.        
remote: Compressing objects: 100% (1043/1043), done.        
remote: Total 43138 (delta 1569), reused 2813 (delta 1501), pack-reused 40176        
Receiving objects: 100% (43138/43138), 8.55 MiB | 7.53 MiB/s, done.
Resolving deltas: 100% (25547/25547), done.
**error: invalid path 'smos-archive/test_resources/example-export/projects/client/bar_2019-03-24_10:46:51.smos'**
fatal: unable to checkout working tree
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'

File: scripts/make-release-tags.hs
Comment (NorfairKing) [inline scripts/make-release-tags.hs:1]: I'm not sure what happened to these?

File: smos-cursor/src/Smos/Cursor/SmosFileEditor.hs [hunk 1]
```diff
@@ -86,12 +86,20 @@ tryLockSmosFile p = do
...
   allocate
     (tryLockFile (fromAbsFile lockFilePath) Exclusive) -- We will edit the file so we need an exclusive lock
+    
+   
     ( \mfl ->
         forM_ mfl $ \fl -> do
-          ignoringAbsence $ removeFile lockFilePath
+          
+          -- The following line break on windows because windows claims the file is still in use by another application
+          -- Swapping the order fixes the issue
```
Comment (NorfairKing) [inline smos-cursor/src/Smos/Cursor/SmosFileEditor.hs:95]: I'm surprised, but not opposed.

File: smos-data/package.yaml [hunk 2]
```diff
@@ -30,3 +30,5 @@ library:
   - validity-text
   - validity-time
   - yaml
+  - Win32
+  - unix-compat
```
Comment (NorfairKing) [inline smos-data/package.yaml:34]: Why do you need these two for the data package??
It's not even used AFAICT

File: smos-query/package.yaml [hunk 3]
```diff
@@ -37,6 +37,7 @@ library:
   - validity
   - validity-path
   - yaml
+  - ncurses
```
Comment (NorfairKing) [inline smos-query/package.yaml:40]: Is this necessary? I'm surprised.

File: smos/package.yaml [hunk 4]
```diff
@@ -50,11 +50,12 @@ library:
   - text
   - time
   - tz
-  - unix
   - unliftio
   - validity
   - vty
-
+  - vty-crossplatform
+  - vty-windows
```
Comment (NorfairKing) [inline smos/package.yaml:57]: I don't think we can have `vty-windows` on linux so we'd have to make that conditional.

File: smos/src/Smos.hs [hunk 5]
```diff
@@ -11,12 +11,13 @@ module Smos
 where
 
 import Brick.BChan as Brick
-import Brick.Main as Brick
+import Brick.Main as BrickS
```
Comment (NorfairKing) [inline smos/src/Smos.hs:14]: I don't understand this change. Is it accidental?

File: smos/src/Smos/Instance.hs [hunk 6]
```diff
@@ -1,23 +1,31 @@
...
+-- import qualified Graphics.Vty.Platform.Unix.Settings  as Vty(settingInputFd,settingOutputFd,defaultSettings) --
+-- import Graphics.Vty.Platform.Unix.Output.TerminfoBased  (setWindowSize)
 import Smos
 import Smos.Terminal
 import Smos.Types
 
+{-
+
+
 withSmosInstance :: (MonadUnliftIO m) => SmosConfig -> Maybe StartingPath -> (TerminalHandle -> m a) -> m a
```
Comment (NorfairKing) [inline smos/src/Smos/Instance.hs:15]: I think this function is only used for testing.
We could split out a `smos-e2e` package to put it in, then `smos-web-server` can use that and the tests can be in there.

File: smos/src/Smos/Terminal.hs [hunk 7]
```diff
@@ -7,10 +7,10 @@ import Conduit
 import Data.Aeson
 import Data.ByteString (ByteString)
 import Data.Word
-import Graphics.Vty.Output.TerminfoBased (setWindowSize)
-import System.Posix
+-- import System.Posix
 import UnliftIO
 
+{-
```
Comment (NorfairKing) [inline smos/src/Smos/Terminal.hs:13]: Same story here, it can be in `smos-e2e`.

File: smos/src/Smos.hs (hunk 5 above)
Comment (tgoossens) [reply smos/src/Smos.hs:14]: Woops. Yes.  CTRL-S going wrong :)

File: smos-data/package.yaml (hunk 2 above)
Comment (tgoossens) [reply smos-data/package.yaml:34]: Doublechecked and confirmed. I removed both.

File: smos-query/package.yaml (hunk 3 above)
Comment (tgoossens) [reply smos-query/package.yaml:40]: Confirmed. I deleted this.  I was an attempt to fix the following error:
image

However, it turned out the problem is that the curses library cannot be found on my windows machine. So for now, smos-query cannot be compiled.

File: smos/src/Smos/Terminal.hs (hunk 7 above)
Comment (tgoossens) [reply smos/src/Smos/Terminal.hs:13]: Fine for me.

File: smos/package.yaml (hunk 4 above)
Comment (tgoossens) [reply smos/package.yaml:57]: I managed to remove the vty-windows dependency. Now everything seems to work with just vty-crossplatform.

File: scripts/make-release-tags.hs
Comment (tgoossens) [reply scripts/make-release-tags.hs:1]: I have not explicitly edited these files. So I also do not know.

**Comment by NorfairKing** [comment]
I split out a `smos-e2e` package here:

https://github.com/NorfairKing/smos/pull/276

**Comment by NorfairKing** [comment]
Superceded, support is on development

**Pull request closed without merging**